pkg/build/types/definition.go

```diff
@@ -20,7 +20,15 @@ type Definition struct {
 	ImageData  `json:"imageData"`
 	BuildData  Data              `json:"buildData"`
 	CustomData map[string]string `json:"customData"`
-	Raw        []byte            `json:"raw"`
+
+	// StageRaw contains the raw data for a particular stage of a multistage
+	// definition file, which amounts to the entire definition file if there is
+	// only one build stage.
+	StageRaw []byte `json:"stageraw"`
+
+	// FullRaw contains the raw data for the entire definition.
+	FullRaw []byte `json:"raw"`
```

Comment:
If the content of `FullRaw` is now different to `Raw` before, then I'd prefer the serialized name changes to `fullRaw` too, as an indicator they aren't the same.

Do we output the full serialization of `Definition` anywhere, or is it just used in tests?

Reply:
This was done in the interest of backward compatibility - since, for single-stage def files, the content of `FullRaw` _is_ the same as what the content of `Raw` was before, and there is still code in the remote-build area that treats all def files as single-stage even if they're not. By which I mean: code that calls parser.ParseDefinitionFile() instead of calling parser.All() and letting the latter call the former as necessary.

An example is: cmd/internal/cli/build.go:445

```go
	if isValid {
		sylog.Debugf("Found valid definition: %s\n", spec)
		// File exists and contains valid definition
		var defFile *os.File
		defFile, err = os.Open(spec)
		if err != nil {
			return types.Definition{}, err
		}

		defer defFile.Close()

		return parser.ParseDefinitionFile(defFile)
	}
``` 

So I was aiming for as much backward-compatibility as we could muster. This is not a concern with the names of the struct fields themselves, because the compiler and linter will tell you if there's still a piece of code floating around trying to access the now-nonexistent `.Raw` field. But there's no such automatic checking of extant JSON files that might be getting used somewhere. (This may be an unwarranted concern, but I don't yet have a grasp of the totality of the system sufficient to assure myself that it's definitely _not_ the case.)

Reply:
I see where you are going here, but:

- If I'm not completely misunderstanding things, it is *not* the same now for multi-stage.... so this is a breaking change in the meaning of a field in a public JSON structure, in a subset of its usage.
- This is targeting a major release, so breaking changes are okay but...
- Changing the field name makes it much more obvious to anyone consuming the JSON that there has been a breaking change they need to consider.

On that basis I'd still prefer to change the name in the JSON. @EmmEff might agree or disagree, so let's hold on it until he has given input.

Reply:
Discussed at length and addressed out of band :-)